### PR TITLE
[GH-727] Block validation

### LINKED ISSUE #727: Adjust block validation to work with key/micro blocks

### PR DESCRIPTION
**Pull request by d-velev**
closes #727 
A check for block gas limit (tx limit) should be added as well when that is done

### CODE REVIEW

File: apps/aecore/lib/aecore/chain/worker.ex [hunk 1]
```diff
@@ -224,13 +242,16 @@ defmodule Aecore.Chain.Worker do
     end
   end
 
-  @spec add_block(Block.t()) :: :ok | {:error, String.t()}
-  def add_block(%Block{header: %Header{prev_hash: prev_hash}} = block) do
+  @spec add_block(Block.t(), boolean()) :: :ok | {:error, String.t()}
```
Comment (grzegorz225) [inline apps/aecore/lib/aecore/chain/worker.ex]: typespec

File: apps/aecore/lib/aecore/util/header.ex [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  @moduledoc """
+  Header utilities
+  """
+  alias Aecore.Chain.KeyHeader
+  alias Aecore.Chain.MicroHeader
+
+  @spec hash(KeyHeader.t() | MicroHeader.t()) :: binary()
+  def hash(header) do
+    case header do
+      %KeyHeader{} ->
```
Comment (grzegorz225) [inline apps/aecore/lib/aecore/util/header.ex]: Code reduction

**Review by gorbak25 — CHANGES_REQUESTED** [review]
Good work but a small compatibility change is required.

File: apps/aecore/lib/aecore/chain/chainstate.ex [hunk 3]
```diff
@@ -64,27 +63,29 @@ defmodule Aecore.Chain.Chainstate do
...
-          {:error, reason} ->
-            {:halt, reason}
-        end
-      end)
-
-    final_chainstate =
-      calculate_miner_reward_chain_state(txs, updated_chainstate, block_height, miner)
+      case block do
+        %KeyBlock{header: %KeyHeader{}} ->
+          chainstate_with_pruned_calls
```
[the remark this answers is not in the document]
Comment (d-velev) [reply apps/aecore/lib/aecore/chain/chainstate.ex]: Yes, #736

File: apps/aecore/lib/aecore/util/header.ex (hunk 2 above)
Comment (d-velev) [reply apps/aecore/lib/aecore/util/header.ex]: This change is present in master-ng, I forgot to set that as the merge target

File: apps/aecore/lib/aecore/chain/block_validation.ex [hunk 4]
```diff
@@ -3,115 +3,156 @@ defmodule Aecore.Chain.BlockValidation do
...
 
-      !is_target_met?(header) ->
-        {:error, "#{__MODULE__}: Header hash doesn't meet the target"}
+      !is_signature_valid ->
+        {:error, "#{__MODULE__}: Invalid micro block signature"}
+
+      txs_hash != calculate_txs_hash(txs) ->
+        {:error, "#{__MODULE__}: Root hash of transactions does not match the one in header"}
+
+      !(new_block |> validate_block_transactions() |> Enum.all?()) ->
```
[the remark this answers is not in the document]
Comment (d-velev) [reply apps/aecore/lib/aecore/chain/block_validation.ex]: Now that I think about it this check isn't required at all, we validate transactions before applying them in Chainstate.calculate_and_validate_chain_state and it does break on the first invalid transaction

File: apps/aecore/lib/aecore/chain/block_validation.ex [hunk 5]
```diff
@@ -149,21 +190,33 @@ defmodule Aecore.Chain.BlockValidation do
...
+    new_prev_key_hash == Header.top_key_block_hash(prev_header)
+  end
+
   @spec check_correct_height?(Block.t(), Block.t()) :: boolean()
-  defp check_correct_height?(%Block{header: %Header{height: new_block_height}}, %Block{
-         header: %Header{height: previous_block_height}
+  defp check_correct_height?(%{header: %{height: new_block_height}}, %{
+         header: %{height: previous_block_height}
        }) do
     previous_block_height + 1 == new_block_height
```
[the remark this answers is not in the document]
Comment (d-velev) [reply apps/aecore/lib/aecore/chain/block_validation.ex]: Good catch

**Comment by d-velev** [comment]
Note: calls and oracle objects are also now pruned only when processing key blocks

**Review by thepiwo — COMMENTED** [review]
comments about refactoring in https://github.com/aeternity/elixir-node/pull/798 apply here

**Review by thepiwo — APPROVED** [review]
looks okay, is this supposed to be merged, with failing tests?

**Comment by d-velev** [comment]
@thepiwo yes

File: apps/aecore/lib/aecore/miner/worker.ex [hunk 6]
```diff
@@ -321,7 +319,7 @@ defmodule Aecore.Miner.Worker do
     txs_list = get_pool_values()
     ordered_txs_list = Enum.sort(txs_list, fn tx1, tx2 -> tx1.data.nonce < tx2.data.nonce end)
 
-    candidate_height = top_height
+    candidate_height = Chain.top_height()
```
Comment (grzegorz225) [inline apps/aecore/lib/aecore/miner/worker.ex]: What if we want to mine on top of an old block(we just now received a malicious block)?

**Review by gorbak25 — APPROVED** [review]

**Comment by thepiwo** [comment]
did somebody else have a look here? @DanielaIvanova @Artur64 @cheezus1

**Review by DanielaIvanova — APPROVED** [review]

**Review by Artur64 — APPROVED** [review]

**Pull request merged**